### PR TITLE
JSUI-2686 Add clear to simple filter

### PR DESCRIPTION
**Pull request by rdeandaCoveo**
Add a clear button to SimpleFilter
Add analytics event to log the clearing of a SimpleFilter
Add unit tests for clear functionality
Add a loop to test all the previous existing tests with different enableClearButton combinations

https://coveord.atlassian.net/browse/JSUI-2686

### CODE REVIEW

File: sass/_SimpleFilter.scss [hunk 1]
```diff
@@ -166,8 +169,22 @@
...
-  height: 6px;
+  width: 12px;
+  height: 8px;
+}
+
+.coveo-simplefilter-eraser {
+  position: relative;
+  top: 1px;
+  margin: 0 5px 0 10px;
+  cursor: pointer;
```
Comment (olamothe) [inline sass/_SimpleFilter.scss]: I think this is already included in the `clearButton` mixin

File: sass/_SimpleFilter.scss [hunk 2]
```diff
@@ -166,8 +169,22 @@
...
+  pointer-events: auto;
```
Comment (olamothe) [inline sass/_SimpleFilter.scss]: Why is this needed ?

File: src/ui/SimpleFilter/SimpleFilter.ts [hunk 3]
```diff
@@ -458,6 +471,23 @@ export class SimpleFilter extends Component {
...
+    this.clearElement.hide();
+
+    new AccessibleButton()
+      .withElement(this.clearElement.el)
+      .withLabel(l('Clear', this.options.title))
+      .withClickAction((evt: Event) => {
+        evt.stopPropagation();
+        this.handleClear();
+      })
+      .withEnterKeyboardAction(() => this.handleClear())
```
Comment (olamothe) [inline src/ui/SimpleFilter/SimpleFilter.ts]: Since you are using an actual `button` element, these are superfluous. The `AccessibleButton` class is mostly meant to transform "non button" element (like, a div or a span) to act like a normal button should (click == enter on keyboard, label, etc).

We have this class in place for backward compatibility reason, where we can't go and change an existing element (that was a div) to a button, so we bridge the gap like this.

In this case, since it's a new option/new markup, I would not use this class at all, and just create a normal button with all correct properties (the label), and it will be accessible by default.

File: unitTests/ui/SimpleFilterTest.ts [hunk 4]
```diff
@@ -1,227 +1,327 @@
...
-        availableSorts: ['score'],
-        sortCriteria: 'score',
-        valueCaption: {
-          bar: 'baz',
-          gmailmessage: 'Gmail Message',
-          lithiummessage: 'Lithium Message',
-          youtubevideo: 'Youtube Video',
-          message: 'Message'
-        }
+    enableClearButtonOptions.forEach(enableClearButton => {
```
Comment (olamothe) [inline unitTests/ui/SimpleFilterTest.ts]: It's kind of weird that we are testing all component `options` inside that loop, but that loop is only meant to tests variation on `enableClearButton`.

What if we want to add a new test for different variation on another option ? We'd have to do some weird changes.

I'd change it so that we have a `describe` block that checks for all options individually, and then inside that block, add another `describe` block for the variations on `enableClearButtonOptions `

File: src/ui/SimpleFilter/SimpleFilter.ts [hunk 5]
```diff
@@ -159,13 +160,20 @@ export class SimpleFilter extends Component {
           return 'score';
         }
       }
-    })
+    }),
+    /**
+     * Specifies whether to show a button to clear all selected values.
```
Comment (fbeaudoincoveo) [inline src/ui/SimpleFilter/SimpleFilter.ts]: Specifies whether to (...)  -> Whether to (...)

File: src/ui/SimpleFilter/SimpleFilter.ts [hunk 6]
```diff
@@ -159,13 +160,20 @@ export class SimpleFilter extends Component {
...
+     *
+     * Default value is `false`.
```
Comment (fbeaudoincoveo) [inline src/ui/SimpleFilter/SimpleFilter.ts]: You can remove this line; TypeDoc will automatically generate a tag indicating what the default value is.

File: unitTests/ui/SimpleFilterTest.ts [hunk 7]
```diff
@@ -1,227 +1,327 @@
...
-      expect(aSimpleFilter.cmp.getValueCaption('bar')).toBe('baz');
-    });
-
-    it('should set the values correctly', () => {
-      expect(aSimpleFilter.cmp.options.values).toEqual(['foo', 'bar']);
-    });
-
-    it('should set the sortCriteria correctly', () => {
-      expect(aSimpleFilter.cmp.options.sortCriteria).toBe('score');
+      describe(`SimpleFilter with more than one instance and enabledClearButton as ${enableClearButton}`, () => {
```
Comment (olamothe) [inline unitTests/ui/SimpleFilterTest.ts]: Same kind of comment here, with that changes it makes it so it will be hard to modify that test file further to test other variation.

Right now it is structured a bit like if the `enableClearButton` was somehow a more "important" option in the component than all the others and that we have to run all unit tests with a variation on that option, and that does not seem right.

File: unitTests/ui/SimpleFilterTest.ts [hunk 8]
```diff
@@ -1,227 +1,327 @@
 import * as Mock from '../MockEnvironment';
```
Comment (olamothe) [inline unitTests/ui/SimpleFilterTest.ts]: There's a lot of changes in that file, would you mind explaining the reasoning/logic/structure you are trying to achieve ? 

It's otherwise a bit hard to follow all the changes.

File: sass/_SimpleFilter.scss (hunk 1 above)
Comment (rdeandaCoveo) [reply sass/_SimpleFilter.scss]: Nope. There aren't . I took those properties from `.coveo-facet-header-eraser` inside `_FacetHeader.scss`.  Maybe we could create something like a `baseHeaderButton` mixin to reuse more properties.

File: sass/_SimpleFilter.scss (hunk 2 above)
Comment (rdeandaCoveo) [reply sass/_SimpleFilter.scss]: I spent quit a few time figuring out why the clear button wasn't triggering the click event, an it was because the `.coveo-simplefilter-select` has `pointer-events: none;` which gets inherited to the children. So `pointer-events: auto;` in a child element allow the click event to be properly triggered.

File: unitTests/ui/SimpleFilterTest.ts (hunk 8 above)
Comment (rdeandaCoveo) [reply unitTests/ui/SimpleFilterTest.ts]: Besides adding specific tests for the clear button behaviour, I wanted to make sure that all the previous existing tests were ok even if enableClearButton was true/false/null. Something like a 'Parameterized test' (https://medium.com/@nyablk97/parameterized-tests-with-jasmine-ecadb2856980) to avoid DRY. In this context, the parameters I used were just an array of booleans, but they could be an array of configuration options with expected results.

But maybe an overkill for the moment.

**Comment by ThibodeauJF** [comment]
See if you can use the `DynamicFacetHeaderButton` class. I added options when you create it in order to use an svg icon instead of simply text. It could save some code.

**Comment by rdeandaCoveo** [comment]
> See if you can use the `DynamicFacetHeaderButton` class. I added options when you create it in order to use an svg icon instead of simply text. It could save some code.

It's just that the name could be more generic to make sense using it outside of a `DynamicFacetHeader`. Maybe `HeaderButton`?

File: src/ui/SimpleFilter/SimpleFilter.ts (hunk 3 above)
Comment (rdeandaCoveo) [reply src/ui/SimpleFilter/SimpleFilter.ts]: Cool

File: unitTests/ui/SimpleFilterTest.ts (hunk 8 above)
Comment (rdeandaCoveo) [reply unitTests/ui/SimpleFilterTest.ts]: I've committed it with the original structure and just the additional tests. Here it's easier to compare the changes https://github.com/coveo/search-ui/compare/JSUI-2686.

File: src/ui/SimpleFilter/SimpleFilter.ts [hunk 9]
```diff
@@ -458,6 +470,21 @@ export class SimpleFilter extends Component {
     return this.circleElement.el;
   }
 
+  public buildClearElement(): HTMLElement {
```
Comment (samisayegh) [inline src/ui/SimpleFilter/SimpleFilter.ts]: I find methods that set class-wide properties and that also have a return type to be a little messy. It's like combining a getter and a setter into one function. I prefer to use or the other.

Also, does this method need to be `public`?
e.g.

```
private buildClearButton(): Dom {
     const button = ...
     return button;
}
```
or
```
private initClearButton() {
    const button = ...
    this.clearButton = button;
}
```

File: src/ui/SimpleFilter/SimpleFilter.ts [hunk 10]
```diff
@@ -458,6 +470,21 @@ export class SimpleFilter extends Component {
...
+    this.clearElement = $$('button', { className: 'coveo-simplefilter-eraser' }, SVGIcons.icons.mainClear);
+    this.clearElement.hide();
+
+    this.clearElement.setAttribute('aria-label', l('Clear', this.options.title));
+    this.clearElement.setAttribute('title', l('Clear', this.options.title));
```
Comment (samisayegh) [inline src/ui/SimpleFilter/SimpleFilter.ts]: Setting the aria-label and title to the same value results in a screen-reader message that is not very nice. It reads "Clear <title> button clear <title>". The `title` should contain additional info for the user. e.g. "deselect all active values in the filter".

File: src/ui/SimpleFilter/SimpleFilter.ts [hunk 11]
```diff
@@ -458,6 +470,21 @@ export class SimpleFilter extends Component {
...
+
+    this.clearElement.on('click', (evt: Event) => {
```
Comment (samisayegh) [inline src/ui/SimpleFilter/SimpleFilter.ts]: There is a class called `AccessibleButton` that accepts the root element and makes it easy to add click handlers, keypress "enter" handlers so that the button is keyboard accessible too.

File: unitTests/ui/SimpleFilterTest.ts [hunk 12]
```diff
@@ -224,5 +228,90 @@ export function SimpleFilterTest() {
...
+      function clickOnClearButton() {
+        $$(simpleFilterContainer.find('.coveo-simplefilter-eraser')).trigger('click');
+      }
+
+      function selectSomeValues() {
+        simpleFilterWithEraser.cmp.selectValue('foo');
+        simpleFilterWithEraser.cmp.selectValue('bar');
+      }
+
+      function getClearButton() {
```
Comment (samisayegh) [inline unitTests/ui/SimpleFilterTest.ts]: If we hoist this function up one level and adjust the implementation to use `cmp.element`, we could reuse it in the test on line 199?

File: unitTests/ui/SimpleFilterTest.ts [hunk 13]
```diff
@@ -224,5 +228,90 @@ export function SimpleFilterTest() {
...
+        return $$(simpleFilterContainer.find('.coveo-simplefilter-eraser'));
+      }
+
+      beforeEach(() => {
+        simpleFilterWithEraser = Mock.optionsComponentSetup<SimpleFilter, ISimpleFilterOptions>(SimpleFilter, <ISimpleFilterOptions>{
+          ...aSimpleFilter.cmp.options,
+          enableClearButton: true
+        });
+        simpleFilterContainer = $$(simpleFilterWithEraser.cmp.element);
```
Comment (samisayegh) [inline unitTests/ui/SimpleFilterTest.ts]: Following on the above comment, is the `simpleFilterContainer` variable really needed? If so, what about using a function instead of needing to "maintain" a variable?

File: unitTests/ui/SimpleFilterTest.ts [hunk 14]
```diff
@@ -224,5 +228,90 @@ export function SimpleFilterTest() {
...
+
+        clickOnClearButton();
+
+        expect(simpleFilterWithEraser.env.usageAnalytics.logSearchEvent).toHaveBeenCalledWith(
+          jasmine.objectContaining({
+            name: 'clearAll',
+            type: 'simpleFilter'
+          }),
+          jasmine.objectContaining({
+            simpleFilterTitle: simpleFilterWithEraser.cmp.options.title
```
Comment (samisayegh) [inline unitTests/ui/SimpleFilterTest.ts]: Should we also assert that the `field` is included?

File: unitTests/ui/SimpleFilterTest.ts [hunk 15]
```diff
@@ -224,5 +228,90 @@ export function SimpleFilterTest() {
...
+          })
+        );
+      });
+
+      it('should log only one search event after clicking on clear button', () => {
+        selectSomeValues();
+        (simpleFilterWithEraser.env.usageAnalytics.logSearchEvent as jasmine.Spy).calls.reset();
+        clickOnClearButton();
+
+        expect(simpleFilterWithEraser.env.usageAnalytics.logSearchEvent).toHaveBeenCalledTimes(1);
```
Comment (samisayegh) [inline unitTests/ui/SimpleFilterTest.ts]: I generally like to keep one expectation per test too. In this case though, I think it's okay to combine this line with the previous test. My thinking is that when the line is removed, only one test fails, but having two tests is fine too.

File: src/ui/SimpleFilter/SimpleFilter.ts [hunk 16]
```diff
@@ -159,13 +160,19 @@ export class SimpleFilter extends Component {
           return 'score';
         }
       }
-    })
+    }),
+    /**
+     * Whether to show a button to clear all selected values.
+     *
```
[the remark this answers is not in the document]
Comment (rdeandaCoveo) [reply src/ui/SimpleFilter/SimpleFilter.ts]: @fbeaudoincoveo pointed out that wasn't necessary as TypeDoc takes care of that.

File: src/ui/SimpleFilter/SimpleFilter.ts (hunk 11 above)
Comment (rdeandaCoveo) [reply src/ui/SimpleFilter/SimpleFilter.ts]: I was using it but @olamothe pointed out that a standard button with appropriate aria-labels will suffice, as `AccessibleButton` was a workaround when it was not possible to change the markup of existing elements and still accessibility was required.

File: src/ui/SimpleFilter/SimpleFilter.ts [hunk 17]
```diff
@@ -371,12 +381,13 @@ export class SimpleFilter extends Component {
...
       ? analyticsActionCauseList.simpleFilterSelectValue
       : analyticsActionCauseList.simpleFilterDeselectValue;
-    this.usageAnalytics.logSearchEvent<IAnalyticsSimpleFilterMeta>(action, {
-      simpleFilterTitle: this.options.title,
-      simpleFilterSelectedValue: checkbox.label,
-      simpleFilterField: <string>this.options.field
-    });
+
     if (this.shouldTriggerQuery) {
+      this.usageAnalytics.logSearchEvent<IAnalyticsSimpleFilterMeta>(action, {
```
[the remark this answers is not in the document]
Comment (rdeandaCoveo) [reply src/ui/SimpleFilter/SimpleFilter.ts]: I think all field-containing properties are of type IFieldOption to enforce starting with an "@".

File: src/ui/SimpleFilter/SimpleFilter.ts (hunk 11 above)
Comment (samisayegh) [reply src/ui/SimpleFilter/SimpleFilter.ts]: Okay, sounds good, the most important is to ensure that the clear button is keyboard accessible in addition to a mouse click.

File: src/ui/SimpleFilter/SimpleFilter.ts (hunk 10 above)
Comment (rdeandaCoveo) [reply src/ui/SimpleFilter/SimpleFilter.ts]: I took that from `DynamicFacetHeaderButton` so maybe it's needing this adjustment too.

File: unitTests/ui/SimpleFilterTest.ts (hunk 13 above)
Comment (rdeandaCoveo) [reply unitTests/ui/SimpleFilterTest.ts]: I think that helps a bit to declutter how the test look and also a bit of performance because less function calls are required.

File: unitTests/ui/SimpleFilterTest.ts (hunk 14 above)
Comment (rdeandaCoveo) [reply unitTests/ui/SimpleFilterTest.ts]: Yes

File: unitTests/ui/SimpleFilterTest.ts (hunk 13 above)
Comment (rdeandaCoveo) [reply unitTests/ui/SimpleFilterTest.ts]: Cool

File: src/ui/SimpleFilter/SimpleFilter.ts (hunk 11 above)
Comment (rdeandaCoveo) [reply src/ui/SimpleFilter/SimpleFilter.ts]: It is ;-)

**Comment by ThibodeauJF** [comment]
> It's just that the name could be more generic to make sense using it outside of a DynamicFacetHeader. Maybe HeaderButton?

Yes if you think a different name is more appropriate, you can move it and rename it to like FacetHeaderButton

File: src/ui/SimpleFilter/SimpleFilter.ts (hunk 10 above)
Comment (ThibodeauJF) [reply src/ui/SimpleFilter/SimpleFilter.ts]: Weird, Voice Over doesn't do that

File: sass/_SimpleFilter.scss (hunk 2 above)
Comment (olamothe) [reply sass/_SimpleFilter.scss]: 👍

File: src/ui/SimpleFilter/SimpleFilter.ts [hunk 18]
```diff
@@ -458,6 +470,21 @@ export class SimpleFilter extends Component {
     return this.circleElement.el;
   }
 
+  public buildClearElement(): HTMLElement {
+    this.clearElement = $$('button', { className: 'coveo-simplefilter-eraser' }, SVGIcons.icons.mainClear);
+    this.clearElement.hide();
+
+    this.clearElement.setAttribute('aria-label', l('Clear', this.options.title));
```
Comment (olamothe) [inline src/ui/SimpleFilter/SimpleFilter.ts]: These attributes could be set when you create the button.
eg: 

```
 this.clearElement = $$('button', { 
   className: 'coveo-simplefilter-eraser',
   title: l('Clear', this.options.title),
   'aria-label': l('Clear', this.options.title)
}, SVGIcons.icons.mainClear);
```

File: unitTests/ui/SimpleFilterTest.ts [hunk 19]
```diff
@@ -224,5 +228,89 @@ export function SimpleFilterTest() {
...
+      beforeEach(() => {
+        simpleFilterWithEraser = Mock.optionsComponentSetup<SimpleFilter, ISimpleFilterOptions>(SimpleFilter, <ISimpleFilterOptions>{
+          ...aSimpleFilter.cmp.options,
+          enableClearButton: true
+        });
+        simpleFilterWithEraser.cmp.toggleContainer();
+      });
+
+      it('should create a clear button if enableClearButton is true', () => {
+        expect(getClearButton()).not.toBeFalsy();
```
Comment (olamothe) [inline unitTests/ui/SimpleFilterTest.ts]: Or `toBeDefined()` ? Instead of double negative ;)

File: src/ui/SimpleFilter/SimpleFilter.ts (hunk 18 above)
Comment (rdeandaCoveo) [reply src/ui/SimpleFilter/SimpleFilter.ts]: Cool

File: src/ui/SimpleFilter/SimpleFilter.ts (hunk 10 above)
Comment (rdeandaCoveo) [reply src/ui/SimpleFilter/SimpleFilter.ts]: I've just tried Voice Over and if the aria-label is the same as title, it reads the label once. But if they're different then it reads the label and then the title. So it does make sense what @samisayegh pointed out.

Comment (ThibodeauJF) [reply src/ui/SimpleFilter/SimpleFilter.ts]: Good to know!
